Sto tutorial gitlab doc 3079 (#1669)
harness/developer-hub: tutorials/security-tests/gitlab-ci-integration.md

---
title: Set up STO integrations with GitLab CI
description: Launch pipeline builds and scans automatically based on GitLab merge requests.
sidebar_position: 25
---

```mdx-code-block
import create_pipeline from './static/trigger-tutorial-00-create-pipeline.png'
import add_sto_stage from './static/trigger-tutorial-01-create-sto-stage.png'
```

This tutorial shows how you can set up a STO pipeline that runs a build and scans a code repository automatically in response to a Git event. You'll do the following:

> * Add a webhook to the code repo that sends Git-event notifications to the Harness pipeline.
> 
> * Add a [Git trigger](/docs/platform/triggers/triggering-pipelines/) to the pipeline that responds to these notifications. 

Reviewer: ```suggestion
* Add a webhook to the code repo that sends Git-event notifications to the Harness pipeline.
* Add a [Git trigger](/docs/platform/triggers/triggering-pipelines/) to the pipeline that responds to these notifications. 
```

> :::note

Reviewer: ```suggestion
:::info
```

suggest using a different note type so it's more visible.

This tutorial uses a [Python repository in Gitlab](https://gitlab.com/gitsequence090/TestPythonSAST) and [Bandit](https://github.com/PyCQA/bandit), an open-source tool designed to find common security issues in Python code.

You can use this workflow for other types of GitLab repositories. However, the specific scanner and setup for your scan step might be different. 

:::

> ### Before you begin

Reviewer: ```suggestion
## Before you begin
```

To do this tutorial, you need the following:

> * A [GitLab](https://gitlab.com/) account. 
> 
>   * An SSH key to enable pushing updates to your repositories. For more information, go to [Use SSH keys to communicate with GitLab](https://docs.gitlab.com/ee/user/ssh.html) in the Gitlab documentation.
> 
>   * A Personal Access Token to enable pushing API requests. For information about the scopes to enable for your token, go to [GitLab connector settings reference](/docs/platform/connectors/code-repositories/ref-source-repo-provider/git-lab-connector-settings-reference/#passwordpersonal-access-token).
> 
> * A Harness [code repository connector](/docs/category/code-repositories) to your Git service.
> 
> * A Harness [Docker connector](/docs/platform/Connectors/Cloud-providers/ref-cloud-providers/docker-registry-connector-settings-reference)

Reviewer: ```suggestion
* A [GitLab](https://gitlab.com/) account. 
  * An SSH key to enable pushing updates to your repositories. For more information, go to [Use SSH keys to communicate with GitLab](https://docs.gitlab.com/ee/user/ssh.html) in the Gitlab documentation.
  * A Personal Access Token to enable pushing API requests. For information about the scopes to enable for your token, go to [GitLab connector settings reference](/docs/platform/connectors/code-repositories/ref-source-repo-provider/git-lab-connector-settings-reference/#passwordpersonal-access-token).
* A Harness [code repository connector](/docs/category/code-repositories) to your Git service.
* A Harness [Docker connector](/docs/platform/Connectors/Cloud-providers/ref-cloud-providers/docker-registry-connector-settings-reference)
```

In the. 4th bullet, why do you not suggest using a GitLab connector?

> ### Set up the STO pipeline

Reviewer: ```suggestion
## Set up the STO pipeline
```

1. In the Harness UI, click **Security Tests**, then **Pipelines**, then **Create a Pipeline**. 

>    <figure>

Reviewer: Figure annotation is only needed in concepts/references. In steps, you only need a plain image or lightbox with alt text.

   ```mdx-code-block
    <img src={create_pipeline} alt="Create a new STO pipeline." height="40%" width="40%" />
    ```

   <figcaption>Figure 1: Create a new pipeline</figcaption>
   </figure> 

2. Enter a pipeline name and click **Start**. 

3. In the visual editor, add a **Security Tests** stage. 

   <figure>

    ```mdx-code-block
    <img src={add_sto_stage} alt="Create a new STO pipeline." height="40%" width="40%" />
    ```

   <figcaption>Figure 2: Add a <b>Security Tests</b> stage</figcaption>
   </figure> 

4. In **About your Stage**, do the following:

   a. Enter a name.

   b. In **Configure Codebase**, select your codebase connector and enter the repository that you want to scan.  
     
   <figure>

   ![](./static/trigger-tutorial-02-set-up-sto-stage.png)

   <figcaption>Figure 3: Set up the <b>Security Tests</b> stage</figcaption>
   </figure> 

>    c. Click **Set up Stage**. 

Reviewer: The alignment is off here due to the use of a, b, c (see preview).

Use *  or renumber 1, 2, 3 for sublist instead

> 4. In **Infrastructure**, specify your build infrastructure. STO supports Cloud (Linux AMD64) and Kubernetes infrastructures.
> 
> 5. In **Execution**, click **Add Step** and select **Bandit**. 
> 
> 6. In **Configure Bandit**, specify the following:
> 
>    a. Scan Mode = **Orchestration**
> 
>    b. Target name  = ***your-name*-dvpwa**
> 
>    c. Target variaent = **<+codebase.branch>**
> 
> 7. Click **Apply Changes** to return to the pipeline, 
> 
> 8. Click **Run** (top right). In **Run Pipeline**, select **Git Branch** for the build type and **master** for the branch name. Then click **Run Pipeline** to ensure that the pipeline runs successfully as configured. 

Reviewer: ```suggestion
5. In **Infrastructure**, specify your build infrastructure. STO supports Cloud (Linux AMD64) and Kubernetes infrastructures.

6. In **Execution**, click **Add Step** and select **Bandit**. 

7. In **Configure Bandit**, specify the following:

   a. Scan Mode = **Orchestration**

   b. Target name  = ***your-name*-dvpwa**

   c. Target variant = **<+codebase.branch>**

8. Click **Apply Changes** to return to the pipeline, 

9. Click **Run** (top right). In **Run Pipeline**, select **Git Branch** for the build type and **master** for the branch name. Then click **Run Pipeline** to ensure that the pipeline runs successfully as configured. 
```

I believe the use of "a b c" is breaking the list. Maybe use * instead.

> ### Add a trigger to your pipeline

Reviewer: ```suggestion
## Add a trigger to your pipeline
```

> In this section, you'll add a trigger to your pipeline that that listsns for incoming requests on a webhook.

Reviewer: ```suggestion
Add a trigger to your pipeline that that listens for incoming requests on a webhook.
```

     <figure>

     ![](./static/trigger-tutorial-05-trigger-config.png)

     <figcaption>Figure 5: Configure the trigger</figcaption>
     </figure> 
     
4. In **Conditions**, set the conditions for running the pipeline. The typical setup is to trigger a build based on the main or master branch of the repo.

   <figure>

   ![](./static/trigger-tutorial-06-trigger-condition.png)

   <figcaption>Figure 6: Set the trigger conditions</figcaption>
   </figure> 
   

5. In **Pipeline Input**, select **Git Branch** for the build type and enter **<+trigger.branch>** for the branch name. Then click **Create Trigger**. 

   <figure>

   ![](./static/trigger-tutorial-08-pipeline-input.png)

   <figcaption>Figure 7: Specify the pipeline input</figcaption>
   </figure> 

   The new trigger now appears in the Triggers table. Note that this table includes a **Webhook** column with a link.

6. Click the **Webhook** link to copy the webhook URL for this trigger to the clipboard. You will now add this webhook to your Git service. 

   <figure>

   ![](./static/trigger-tutorial-07-copy-url-webhook.png)

   <figcaption>Figure 8: Copy the webhook URL</figcaption>
   </figure> 

> ### Add the trigger webhook to your Git service

Reviewer: ```suggestion
## Add the trigger webhook to your Git service
```

The steps to do this differ depending on the service you're using. These steps assume you're using GitLab.

1. Go to your GitLab project and choode **Settings** > **Webhooks** in the left menu. 

2. Enter the webhook URL and your GitLab personal access token. 

3. Select the **Merge request event** as the event type to trigger the webhook. 

4. Click **Add webhook** at the bottom of the page. 

> ### Create a Merge Request in GitLab

Reviewer: ```suggestion
## Create a Merge Request in GitLab
```

> At this point, your setup should be complete. To verify that your trigger is working correctly, do the following:

Reviewer: ```suggestion
To verify that your trigger is working correctly, do the following:
```

"your setup should be complete" kind of makes it sound like the tutorial author doesn't trust their tutorial.

> * Open a browser page and point it to the **Execution History** page for your Harness pipeline. 
> 
> * Go to your GitLab repo, then create and approve a merge request. 

Reviewer: ```suggestion
1. Open a browser page and point it to the **Execution History** page for your Harness pipeline. 
2. Go to your GitLab repo, then create and approve a merge request. 
```
